src/H5Dmpio.c

```diff
@@ -1173,26 +1163,40 @@ H5D__piece_io(H5D_io_info_t *io_info)
                 }
             }
         }
+    }
 
-        /* Perform multi dataset I/O if appropriate */
-        if (use_multi_dset) {
+    /* Perform multi dataset I/O if appropriate */
+    if (use_multi_dset) {
 #ifdef H5_HAVE_INSTRUMENTED_LIBRARY
-            /*** Set collective chunk user-input optimization API. ***/
-            if (H5D_ONE_LINK_CHUNK_IO == io_option) {
-                if (H5CX_test_set_mpio_coll_chunk_link_hard(0) < 0)
-                    HGOTO_ERROR(H5E_DATASET, H5E_CANTSET, FAIL, "unable to set property value");
-            } /* end if */
-#endif        /* H5_HAVE_INSTRUMENTED_LIBRARY */
+        /*** Set collective chunk user-input optimization API. ***/
+        if (H5D_ONE_LINK_CHUNK_IO == io_option) {
+            if (H5CX_test_set_mpio_coll_chunk_link_hard(0) < 0)
+                HGOTO_ERROR(H5E_DATASET, H5E_CANTSET, FAIL, "unable to set property value");
+        } /* end if */
+#endif    /* H5_HAVE_INSTRUMENTED_LIBRARY */
+
+        /* Process all the filtered datasets first */
+        if (io_info->filtered_count > 0) {
+            if (H5D__link_chunk_filtered_collective_io(io_info, io_info->dsets_info, io_info->count, mpi_rank,
+                                                       mpi_size) < 0)
+                HGOTO_ERROR(H5E_IO, (H5D_IO_OP_READ == io_info->op_type ? H5E_READERROR : H5E_WRITEERROR),
+                            FAIL, "couldn't finish filtered linked chunk MPI-IO");
+        }
 
+        /* Process all the unfiltered datasets */
+        if ((io_info->filtered_count == 0) || (io_info->filtered_count < io_info->count)) {
             /* Perform unfiltered link chunk collective IO */
             if (H5D__link_piece_collective_io(io_info, mpi_rank) < 0)
-                HGOTO_ERROR(H5E_IO, H5E_CANTGET, FAIL, "couldn't finish linked chunk MPI-IO");
+                HGOTO_ERROR(H5E_IO, (H5D_IO_OP_READ == io_info->op_type ? H5E_READERROR : H5E_WRITEERROR),
+                            FAIL, "couldn't finish linked chunk MPI-IO");
         }
     }
-
-    if (!use_multi_dset) {
+    else {
         /* Loop over datasets */
         for (i = 0; i < io_info->count; i++) {
+            if (io_info->dsets_info[i].skip_io)
+                continue;
```

Comment:
Don't try to perform I/O on datasets where we're skipping I/O, as certain fields aren't initialized and will cause crashes.